### PR TITLE
BorderBoxControl: Promote to stable

### PR DESCRIPTION
**Pull request by DaniGuardiola**
## What?

Part of https://github.com/WordPress/gutenberg/issues/59418

Promote `BorderBoxControl` to a non-experimental component.

## How?

- [x] Export it from components without the `__experimental` prefix.
- [x] Switch subcomponents to dot notation if applicable.
- [x] Keep the old `__experimental` export for backwards compatibility.
- [x] Remove experimental callout from readme.
- [x] Change all imports of the old `__experimental` in GB and components to the one without the prefix (including in Storybook stories).
- [x] Update the docs (readme, JSDocs, Storybook) to refer to the new unprefixed component.
- [x] Add the component storybook `id` (get it from the storybook URL) to the `PREVIOUSLY_EXPERIMENTAL_COMPONENTS` const array in `manager-head.html` so that old experimental story paths are redirected to the new one.
- [x] Add a changelog for both the promotion (enhancement) and the deprecation.

Docs tweaked slightly. No unstable props (other than the 40px sizes one, and `__experimentalIsRenderedInSidebar` which is tracked in https://github.com/WordPress/gutenberg/issues/65581). If no one has any concerns, this one should be a straightforward merge.

### CODE REVIEW

File: packages/components/src/border-box-control/stories/index.story.tsx
Comment (aaronrobertshaw) [inline packages/components/src/border-box-control/stories/index.story.tsx]:
```suggestion
	title: 'Components/BorderBoxControl',
```

Typo?

Edit: It appears that this also causes the component styles in the Storybook to break.

File: packages/components/src/border-box-control/border-box-control/component.tsx
Comment (aaronrobertshaw) [inline packages/components/src/border-box-control/border-box-control/component.tsx:162]: Could the behaviour previously described here be unexpected at all? What's the thinking behind removing this additional information and context?

Is the reasoning for removing this just to avoid a wordy description in IDE hints etc? In that case, would it be better to move this information to the readme?

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for putting together these stabilization PRs @DaniGuardiola 👍 

I've given this a test and it is pretty close. Within the block and site editors the component works as expected. I did run into a minor issue with the Stylebook examples though.

✅ Component unit tests pass
✅ Storybook redirect works as advertised
❓ There appears to be a typo in the storybook title
❓ The component's styles appear messed up however fixing the typo fixes them up.
✅ Didn't spot any regressions in the post/site editors' uses of the component
❓ Left a question around the docs changes but it doesn't have to be a blocker
✅ Consistent with BorderControl stabilization approach settle on in https://github.com/WordPress/gutenberg/pull/65475


| LInked | Unlinked |
|---|---|
| Screenshot 2024-09-24 at 4 27 54 pm | Screenshot 2024-09-24 at 4 28 26 pm |

File: packages/components/src/border-box-control/stories/index.story.tsx
Comment (DaniGuardiola) [reply packages/components/src/border-box-control/stories/index.story.tsx]: Oops! Good catch!

File: packages/components/src/border-box-control/border-box-control/component.tsx
Comment (DaniGuardiola) [reply packages/components/src/border-box-control/border-box-control/component.tsx:162]: If you look at the README (see https://github.com/WordPress/gutenberg/blob/b7f9af87f212b83c68d6f1b92daa6d9ee1e2024d/packages/components/src/border-box-control/border-box-control/README.md), you'll see that there's a general description and then a "Development guidelines" section. This component breaks from the norm by having the "development guidelines" as JSDoc instead of the main description of the component.

My change here is just to bring consistency with what we normally do. Developers will still be able to read those "development guidelines" where they are used to, like with any other components.

Comment (aaronrobertshaw) [reply packages/components/src/border-box-control/border-box-control/component.tsx:162]: 🤦 Yep. Sorry, I thought I checked but comparing this and the other BorderControl PR I probably glanced at the wrong readme. Ignore me then!

File: packages/components/src/border-box-control/stories/index.story.tsx
Comment (tyxla) [inline packages/components/src/border-box-control/stories/index.story.tsx:19]: I think we need to add this:

```diff
diff --git a/packages/components/src/border-box-control/stories/index.story.tsx b/packages/components/src/border-box-control/stories/index.story.tsx
index 7ac8cbbea8c..5341dacab64 100644
--- a/packages/components/src/border-box-control/stories/index.story.tsx
+++ b/packages/components/src/border-box-control/stories/index.story.tsx
@@ -83,4 +83,5 @@ export const Default = Template.bind( {} );
 Default.args = {
        colors,
        label: 'Borders',
+       enableStyle: true,
 };
```

without it, Storybook will display the control as `false` by default:


https://github.com/user-attachments/assets/d956b80b-0423-445c-879f-3373846a19f3

File: packages/components/src/index.ts
Comment (tyxla) [inline packages/components/src/index.ts:41]: What about those experimental utils? They are part of the component's API, but they remain experimental. Should we stabilize them as well?

**Review by tyxla — COMMENTED** [review]
This looks good and tests well 👍 

I'm a bit torn about the `__experimentalIsRenderedInSidebar` prop. As an external consumer, I find it consuming. Even if this prop exists, I think it should have a different name that reflects the behavior of the underlying component. IIRC it was specifying some alternative popover props under the hood. Why does that have to be sidebar-specific behavior? Feels like the sidebar shouldn't have to do anything to do with `BorderBoxControl` at all. 

I'm also concerned that the component continues to export experimental util APIs, which we're not stabilizing here. It feels like we're doing half the work and not really assessing the stabilization of all the APIs of this component.

There's also a minor Storybook thing, I've left a separate comment about it.

File: packages/components/src/index.ts
Comment (DaniGuardiola) [reply packages/components/src/index.ts:41]: This is something I want to address in a follow-up. I already gave it a few iterations some time ago, and even paired on it with Marcelo over a donut chat which resulted in a draft PR I plan on reviving. There was some consensus about not exporting these from the component package, though the details remain to be discussed and iterated on.

I don't think it should block stabilization of this component though.

File: packages/components/src/border-box-control/stories/index.story.tsx
Comment (DaniGuardiola) [reply packages/components/src/border-box-control/stories/index.story.tsx:19]: Yeah, that's a good idea. Better default 👍

File: packages/components/src/index.ts
Comment (tyxla) [reply packages/components/src/index.ts:41]: > I don't think it should block stabilization of this component though.

Why, though, aren't those part of the component's public API? I'd expect it would be the other way around, and we'd figure these out first, especially if we plan not to export them from the components package / deprecate them.

**Comment by DaniGuardiola** [comment]
We need to decide how we want to proceed regarding `__experimentalIsRenderedInSidebar`. In https://github.com/WordPress/gutenberg/pull/65475 we seemed to have found some agreement about it not blocking stabilization of components, and I created https://github.com/WordPress/gutenberg/issues/65581 to track that separately.

Here, @tyxla seems to be arguing against it. That's fine, I probably even agree with it though I generally do not have any strong feelings about either approach. However, we should make a decision once and apply it consistently across components, especially since this prop is shared by a few of them. What do you think? cc @WordPress/gutenberg-components

File: packages/components/src/index.ts
Comment (DaniGuardiola) [reply packages/components/src/index.ts:41]: Related: https://github.com/WordPress/gutenberg/issues/61135

**Comment by tyxla** [comment]
> We need to decide how we want to proceed regarding `__experimentalIsRenderedInSidebar`. In #65475 we seemed to have found some agreement about it not blocking stabilization of components, and I created #65581 to track that separately.
> 
> Here, @tyxla seems to be arguing against it. That's fine, I probably even agree to it though I generally do not have any strong feelings about either approach. However, we should make a decision once and apply it consistently across components, especially since this prop is shared by a few of them. What do you think? cc @WordPress/gutenberg-components

Definitely don't intend to block the PR with this - I also acknowledge the prop is used across multiple components and it might make more sense to be addressed separately. Was just sharing my $0.02, but don't consider this a blocking comment. Maybe if we take care of #65581 not long after stabilizing the component, it should be OK.

File: packages/components/src/index.ts
Comment (DaniGuardiola) [reply packages/components/src/index.ts:41]: > Why, though, aren't those part of the component's public API? I'd expect it would be the other way around, and we'd figure these out first, especially if we plan not to export them from the components package / deprecate them.

I'm not strongly opposed that that, but to answer your question, my opinion is: no, they are not part of the component's API. I think these utils shouldn't have been exported here in the first place, and it's just an odd occurrence of misplaced utils. They're definitely _related_ to the component, in that they apply to borders, but the component is its own thing.

So, the way I see it: stabilizing the component, moving the utilities somewhere else - these are separate tasks. Should they be tackled somewhat close to each other? I think so. Should one block the other? I think not.

That's just my 2 cents though. I'm good to go with another approach if someone feels more strongly about it.

Comment (tyxla) [reply packages/components/src/index.ts:41]: Fair enough. Let's keep the utils a separate problem and solve it as part of #61135 👍

**Review by tyxla — APPROVED** [review]
Considering that we'll take care of the experimental utils in #61135 and the experimental props in #65581, this looks and works well and is good to go 🚀

**Pull request merged**
